types/react/test/tsx.tsx

```diff
@@ -392,6 +393,8 @@ const divRef = React.createRef<HTMLDivElement>();
 <ForwardRef2 ref={divRef}/>;
 // @ts-expect-error
 <ForwardRef2 ref='string'/>;
+// @ts-expect-error
+<ForwardRef2 ref={badlyAuthoredRef} />;
```

Comment:
Why shouldn't `{ current: undefined | null }` be passed to the component instance?

Is this something we want to prevent, or a workaround for the way the current code is written?

Reply:
Also, happy to move this to a different issue/thread.

Reply:
> It seems fine to me to include undefined in that union, as long as null is also there (since React sets it to null).

And it is. I'd love to get this fixed but this is not up to type declarations but how TypeScript works.

 At a high level, TypeScript always assumes that an object passed along is read by whoever sees it. However, React typings currently expect `T | null` so TypeScript thinks that passing `{ current: undefined }` would break it. At runtime, React will only ever write the `current` so it doesn't really care what you passed along i.e. you could pass `T | number` and React would be fine. React only cares about that you're prepared for `T | null`.

In TS jargon: object properties are covariant. What we actually want is contravariant though. Function parameters are already contravariant so we could type callback refs correctly. We just don't do it yet (see the "bivariance hack") so you don't run into any type-inconsistencies when switching between callback refs and ref objects.

I tried to illustrate this in a [TypeScript playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcA5FDvmQFA3AB2MWUBeWmWBAPACoB8cAN404YuLgCuUSkwBccXjQC+dGAE8wHbAQDCyADYGARngDWfQQF44ACikys8xQEo4VwQDcIwACZ1GZlZ2OF1JVBgSACFJGEiGAFEDLBAneCwAD2YGX3QACV4AWQAZGLiIROTUpmFRcRgACxSsBQioRgBzFToaXyxcA2RKCQqIiXDIkDL4nQUdbjCI6Nj4pOamfl7+weHcUfhjFYrZzh4CkumKteqYTb6BoY49hjGsKrST+fPi67S77ceIxe8AYkiMyGMyUuDE+XG432hvxqAB84KCjJsaAB6LFwADqHAA7sAjHAKgZ1HAsJ4WHBKMhfHAAAaUAgAOgcshgTJoBEkDHwwAqnAZAHkGBSdLYhKzlAoZVx5Zy0goEUdKusYMplG4ROIRfg2eEsAkCAR+jBbLY3B5avr9c8xodygx3HSuBzpFyANx0e3iZ3xNmDYxvVBsgjQBJ4Bq2ENvdyCPX++2OiDJYMQDpxiFvFx1f06gtwIvFygwaSu9EGbrY3EAMUYHGQ8Cg-JgoCwbLgCQyeHgGi0cBALdwDRo3GwYol6h07oIViEgeOXGUcCx-F9OLgUQZKFb7c73YJcGJpPpjKZb01wdzBlQTNPDWAY+HwA6DRBEHSGWAY2FjR-lS7xMGyE5Tr44qSlw86LteNw6GuG5briADKJBYIBDAdMO0AcKgWi4MABAvqe0BmOGcAAJoQJIEjIK6hiEsg6joGAaDoKgGFYThxKNLhwxgMQYDoJEcBcakPFwI0LZkjOcB9IY6B8Q0KDiZIxioJhZIEHAQkQCJYGTjgkEznOrKLlISxTOqiHrpudDbrujKyW2TBHiKpnQQQZDoKAYAgTALZCq6vgQFg6AMN+cANAxvjJMyrKeo4TCPqGnRomCNbGdO3mwUI1YQlCtmrvZvp1viRIkgYckUlSNJQKe7TMIlHrKqlvL8oKwp4s1WBQbOXDSrK8qskqXoqnAaouki8CotW2q6sWU6GsaprmvgVo2kmxb6kl7XwDYYVSDcbIAI6SCw6iode+DQLYZDLgwZD5im7oVlArp2NtdpveI+0TTUNjVrt4iqPqpZ7ZhlaZUYtbbo2DDNge7mpN2vb9tJmgcCOMBjhOvXAMwA3mVwi5PXZyEVYjyN0oeaOVSgBjMaxTVE-h3HPthAn4YRxGkYwMUsF2cDoU+4UNe6DLoCybWA9y4kNLRBiMrFNJohAcB-qgl3hgTfUkzBFlCPBHylVTTl7q59Miye5KUtStKEn1anTasgVa0wmuy+yB1pbE4uO41F4ywDKUK5khEwOgyDMoskyIoFTJGYTxNmUbZNCFZCclQQSEOdTTb7nTqMixBLDoBAuk+8lXKPkxLHoGrHBiR0khDC5HTIIwYw1373ZlMyqf9enBCPvbbPMJRACS8BhRFz2tjgRiUrggKErFA7Y61vvy4+sV69ww+G7pxuFZCWDQpTBd8gKHbCvoRimLgZhSgqeejYqpyPyY5jwkUpR1SzTgPNLK-AdS-TECtGARotLrQtFtRMkD7SslsMdSQp0LpXRuskO6UAHpPReq9FM5YYbfSQcmP685bDVmIYWYskN-rQ0+rDGsqgtgPF2PscY1loQ-2fq-L+Oh+F-3jssGaHtQEYg4TsJ43CnoiJfrCPQhhf4v3-hcIBkjWH-E4XI4EwFNSKMEQQOYXBjEaJ+No6smIEZF2tqXdGfZ8BYyHLjfGCxVECNJguJc6pjHXxQnAGmxc3IdgZieBurNnbs3EpzDKIA8LiT5iRXAnshaUG7GLDeEtaQhx3rXNIj5UBKzBKrZA6sopa1QDrCKRkAkZ18abJgDS85lQqs5UJNtuxMmzuI+IrTHyUAMMACKckCl+zgOlbmTIxE2QkZqR8UV4ASUwlzHCiThgQjaP2epXjzA+MshMfpFRWn519EAA).

I'll add a comment with the [underlying TypeScript issue](https://github.com/microsoft/TypeScript/issues/10717#issuecomment-461704350) explaining why this ref is considered "badly authored". Don't be confused by the issue being closed. The variance annotations added do not change any semantics. They're just a performance hint. The underlying issue still exists in https://github.com/microsoft/TypeScript/issues/21759